### PR TITLE
Upgrade requirements and dependencies to work with PHP 7.2 and ES 6.0

### PR DESCRIPTION
**Pull request by CauanCabral**
Current Elastica version don't support PHP 7.2, the current version of language. Also, there's no support for Elastic >= 6 (current stable).

That PR are just a try to make the package compatible with all current active versions of PHP.

- [x] [Replace repository mapping from types to indexes](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/removal-of-types.html)
- [x] ~[Adopt join datatype to parent/child relations](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/parent-join.html)~ Unnecessary as we use Elastica Documents for that.

### CODE REVIEW

**Comment by markstory** [comment]
Looks like a good start.

**Comment by CauanCabral** [comment]
Hi, I would appreciate some opinion about a thing...

Since version 5.0, the Elastic team (recommends use of just one type mapping for each index)[https://www.elastic.co/blog/index-type-parent-child-join-now-future-in-elasticsearch].

Last major stable version (6.0) had dropped support for multiple type mappings in one index. Just legacy index work with multiple types for now (created before upgrade).

Is reasonable to change class name Type for Index or other, most aligned with current elastic-search terminology?

Also, can I use direct mapping between ODM index <-> ORM table?

**Comment by markstory** [comment]
> Is reasonable to change class name Type for Index or other, most aligned with current elastic-search terminology?

Yes, but that would be a breaking change for this plugin (which is ok in my opinion) as elastic search has had dramatic breaking changes as well.

**Comment by CauanCabral** [comment]
I keeped the name Type but included another property named 'index', to store related elastic search index.

Some of methods for query had been deprecated and/or removed between ES 2.3 and 6.0, so I removed it from QueryBuilder wrapper.

There's something else to do?

**Comment by CauanCabral** [comment]
I did a squash of commits and I think is done.

Anything to improve?

**Comment by lorenzo** [comment]
Could you write something in the readme about using the plugin now?

* How to configure connection details for an index?
* Do w need separate connection details for each thing in `IndexRegistry`?

File: tests/TestCase/IndexRegistryTest.php [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+use Cake\Core\Configure;
+use Cake\Core\Plugin;
+use Cake\ElasticSearch\Index;
+use Cake\ElasticSearch\IndexRegistry;
+use Cake\TestSuite\TestCase;
+
+/**
+ * Used to test correct class is instantiated when using IndexRegistry::get();
+ */
+class MyUsersIndex extends Index
```
Comment (markstory) [inline tests/TestCase/IndexRegistryTest.php]: Could we stick to one class per file?

Comment (CauanCabral) [reply tests/TestCase/IndexRegistryTest.php]: Sure.

**Comment by qqqqb** [comment]
Hello @CauanCabral

I wil have a look on you changes and I will start deploying new way on elasticsearch **parent/child**, **join field** and **supporting single type**

Changes timeline:
- Force single type on settings with allowing multi Entities
- Each Entity will have field to point it on elasticsearch
- Add relationship hasOne, hasMany and belongsTo using join field and parent/chlid relationship
- Marge joins of relationship to be like contain nested entities

**Comment by CauanCabral** [comment]
Hi @qqqqb, thank's.

I will make my lastest changes today, but I think you can base your work in current branch state.
Supporting relationships with join field would be great.

What do you mean with "force single type on settings" and "each entity will have field to point"?
My current suggestion is to have two properties in Index class, one for index name (Index::name) and other for the mapping type (Index::type) - not a settings as the old way (in app.php or global defined).

**Comment by CauanCabral** [comment]
I'm done, if you need some help, just get in touch.

**Comment by qqqqb** [comment]
I took fork on this repository to be able writing some wiki documents for merging upgrades.

What I will do is writing the type name on the settings, and add field called Type_name for each document. This way will return multi types feature on ES6 style :)

Some fields will be mandatory to index any document to helps on joining and relationships

Notes, I will take into consideration elasticsearch future upgrades until ES9.

I started working on it and I copied some of your codes. I will comment on that lines once I finish it all. If I need help I will ask you for sure and thank you in advance

**Comment by qqqqb** [comment]
Hello guys,

Could you help me on testing this updates please
[https://github.com/cakephp/elastic-search/compare/master...qqqqb:master](https://github.com/cakephp/elastic-search/compare/master...qqqqb:master)

You will need [https://github.com/ruflin/Elastica/tree/5.x](https://github.com/ruflin/Elastica/tree/5.x) for ES6
Those updates working find with ^ES2

**Comment by lorenzo** [comment]
@qqqqb open another ticket about it, please. This pull request is for discussing the use of ES6

**Comment by lorenzo** [comment]
I'll merge this and tag a beta release.

Thanks for the great work @CauanCabral !

**Pull request merged**